docs/books/valhalla-mjolnir/src/01_valhalla_tiled_structure.md

```diff
@@ -0,0 +1,147 @@
+
+This hierarchical structure allows the routing algorithm to quickly traverse long distances by using higher-level tiles, then switch to more detailed tiles as it approaches the origin and destination.
+
+## Tile Coordinate System
+
+Valhalla uses a custom tile coordinate system based on the following components:
+
+```cpp
+// From baldr/graphid.h
+struct GraphId {
+  uint64_t value;
+  
+  // Methods to access specific parts of the ID
+  uint32_t tileid() const;
+  uint32_t level() const;
+  uint32_t id() const;
+  
+  // ...
+};
+```
+
+Each `GraphId` contains:
+- **Level**: The hierarchy level (0, 1, or 2)
+- **Tile ID**: Identifies a specific tile within the level
+- **ID**: Identifies a specific element (node or edge) within the tile
+
+The tile ID is derived from the geographic coordinates using a space-filling curve algorithm, which ensures that nearby locations are likely to be in the same tile or adjacent tiles.
+
+## Tile Size and Resolution
+
+The size of tiles varies by level:
+
+- Level 0 tiles are the largest, covering approximately 4 degrees × 4 degrees
+- Level 1 tiles cover 1 degree × 1 degree
+- Level 2 tiles are the smallest, covering 0.25 degrees × 0.25 degrees
+
+This variable resolution allows for efficient storage and retrieval of road network data at different scales. The world is divided into a grid based on these tile sizes, with rows and columns starting from the bottom left (-180, -90) and increasing to the top right (180, 90).
+
+## Code Example: Tile Coordinates Calculation
+
+Here's how Valhalla converts geographic coordinates to tile coordinates:
+
+```cpp
+// From midgard/tiles.h
+class Tiles {
+public:
+  /**
+   * Get the tile ID given the latitude and longitude.
+   * @param lat Latitude value.
+   * @param lng Longitude value.
+   * @return Returns the tile ID.
+   */
+  int32_t TileId(const float lat, const float lng) const;
+  
+  // ...
+};
+```
+
+The implementation uses a projection system to map geographic coordinates to tile IDs:
+
+```cpp
+// From midgard/tiles.cc
+int32_t Tiles::TileId(const float lat, const float lng) const {
+  // Get the indexes within the tiles and return the tile ID
+  const int32_t x = static_cast<int32_t>(floor((lng - tilebounds_.minx()) * inv_tile_size_));
+  const int32_t y = static_cast<int32_t>(floor((lat - tilebounds_.miny()) * inv_tile_size_));
+  return (y * ncolumns_) + x;
+}
+```
+
+## Tile Storage Format
+
+Valhalla stores tiles as binary files on disk. Each tile contains:
+
+1. **Header**: Contains metadata about the tile
+2. **Nodes**: Represent intersections in the road network
+3. **Directed Edges**: Represent road segments with direction-specific attributes
+4. **Edge Info**: Contains shared data for edges (e.g., shape, names)
+5. **Additional Data**: Restrictions, traffic signs, admin information, etc.
+
+The binary format is designed for efficient storage and quick loading. Here's the structure of a tile header:
```

Comment:
yeah specifically we can memory map it and its very close to a "no-parse" format